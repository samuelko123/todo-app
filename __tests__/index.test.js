import {
	render,
	screen,
	waitFor,
} from '@testing-library/react'
import {within} from '@testing-library/dom'
import userEvent from '@testing-library/user-event'

import App from '../pages/_app'
import Page from '../pages/index'

it('should render correctly', async () => {

	//============================================================
	
	// Arrange
	render(<App Component={Page} />)

	// Assert
	const textbox = screen.getByRole('textbox', { name: 'Enter New Todo' })
	expect(textbox).toHaveFocus()

	//============================================================

	// Arrange
	const todoText = 'Test Todo Item'
	const btnAdd = screen.getByRole('button', { name: 'Add' })

	// Assert - Before
	
	expect(btnAdd).toBeDisabled()

	// Action
	userEvent.type(textbox, todoText)

	// Assert - After
	await waitFor(() => {
		expect(btnAdd).not.toBeDisabled()
	})

	//============================================================

	// Arrange
	const list = screen.getByRole('list')

	// Assert - Before
	expect(list.children.length).toEqual(0)

	// Action
	userEvent.click(btnAdd)

	// Assert - After
	await waitFor(() => {
		// Expect textbox to be cleared
		expect(textbox.textContent).toEqual('')

		// Expect 1 item in TodoList
		expect(list.children.length).toEqual(1)
		expect(list.children[0]).toHaveTextContent(todoText)
	})

	//============================================================

	// Arrange
	const checkbox = screen.getByRole('checkbox', 'Complete Todo')
	const listItem = screen.getByText(todoText)

	// Assert - Before
	expect(checkbox).not.toBeChecked()
	expect(listItem).toHaveStyle('text-decoration: none')

	// Action #1
	userEvent.click(checkbox)

	// Assert #1
	await waitFor(() => {
		expect(checkbox).toBeChecked()
		expect(listItem).toHaveStyle('text-decoration: line-through')
	})

	// Action #2
	userEvent.click(checkbox)

	// Assert #2
	await waitFor(() => {
		expect(checkbox).not.toBeChecked()
		expect(listItem).toHaveStyle('text-decoration: none')
	})

	//============================================================

	// Arrange
	const deleteBtn = screen.getByRole('button', { name: 'Delete' })

	// Assert - Before
	expect(screen.queryByRole('dialog')).not.toBeInTheDocument()

	// Action
	userEvent.click(deleteBtn)

	// Assert - After
	await waitFor(() => {
		expect(screen.getByRole('dialog')).toHaveTextContent('Delete')
	})

	//============================================================

	// Arrange
	const deleteDialog01 = screen.getByRole('dialog')
	const cancelBtn = within(deleteDialog01).getByRole('button', { name: 'Cancel' })

	// Assert - Before
	expect(list.children.length).toEqual(1)

	// Action
	userEvent.click(cancelBtn)

	// Assert - After
	await waitFor(() => {
		expect(screen.queryByRole('dialog')).not.toBeInTheDocument()
		expect(list.children.length).toEqual(1)
	})

	//============================================================

	// Arrange
	userEvent.click(deleteBtn)
	await waitFor(() => {
		expect(screen.getByRole('dialog')).toHaveTextContent('Delete')
	})
	const deleteDialog02 = screen.getByRole('dialog')
	const confirmBtn = within(deleteDialog02).getByRole('button', { name: 'Delete' })

	// Assert - Before
	expect(list.children.length).toEqual(1)

	// Action
	userEvent.click(confirmBtn)

	// Assert - After
	await waitFor(() => {
		expect(screen.queryByRole('dialog')).not.toBeInTheDocument()
		expect(list.children.length).toEqual(0)
	})

	//============================================================
})