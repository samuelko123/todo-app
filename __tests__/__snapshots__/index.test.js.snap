// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.c6 {
  background-color: #0d5257;
  border: 2px solid #0d5257;
  color: #f8f9fa;
  padding: 0.25rem 0.5rem;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
  text-align: center;
  display: inline-block;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  cursor: pointer;
}

.c6:hover:not(:disabled) {
  background-color: #f8f9fa;
  color: #0d5257;
}

.c6:disabled {
  opacity: 0.5;
  cursor: unset;
}

.c4 {
  position: absolute !important;
  width: 1px !important;
  height: 1px !important;
  padding: 0 !important;
  margin: -1px !important;
  overflow: hidden !important;
  -webkit-clip: rect(0,0,0,0) !important;
  clip: rect(0,0,0,0) !important;
  white-space: nowrap !important;
  border: 0 !important;
}

.c0 {
  padding: 1rem;
  background-color: #0d5257;
  color: #f8f9fa;
  font-size: 1.25rem;
}

.c2 {
  display: grid;
  grid-template-columns: 1fr;
  grid-row-gap: 1rem;
  width: 80%;
  max-width: 100%;
  margin: 1rem auto;
}

.c1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
}

.c7 {
  list-style: none;
  padding: 0;
  margin: 0;
}

.c5 {
  width: 100%;
  background-color: #f8f9fa;
  border: 2px solid rgba(0,0,0,0.125);
  color: #215259;
  padding: 0.25rem 0.5rem;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
}

.c5:focus {
  outline: none;
  box-shadow: 0px 0px 0.25rem #0d5257;
}

.c3 {
  display: grid;
  grid-template-columns: 1fr 4rem;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  padding: 0.25rem 0.5rem;
}

<div>
  <header
    class="c0"
  >
    <a
      class="c1"
      href="/"
    >
      Todo App
    </a>
  </header>
  <main
    class="c2"
  >
    <form>
      <div
        class="c3"
      >
        <label>
          <span
            class="c4"
          >
            Enter New Todo
          </span>
          <input
            class="c5"
            name="new-todo"
            placeholder="Enter New Todo"
          />
        </label>
        <button
          class="c6"
          disabled=""
          type="submit"
        >
          Add
        </button>
      </div>
    </form>
    <ul
      class="c7"
    />
  </main>
</div>
`;
